include/insns/Zicsr.h

```diff
@@ -0,0 +1,184 @@
+      } else {
+        return false;  // Hi half is not available on RV64
+      }
+    }
+    R->AdvancePC( Inst );
+    return true;
+  }
+
+  static constexpr auto& rdcycle    = perfCounter<Lo, rdcycleImpl>;
+  static constexpr auto& rdcycleh   = perfCounter<Hi, rdcycleImpl>;
+  static constexpr auto& rdtime     = perfCounter<Lo, rdtimeImpl>;
+  static constexpr auto& rdtimeh    = perfCounter<Hi, rdtimeImpl>;
+  static constexpr auto& rdinstret  = perfCounter<Lo, rdinstretImpl>;
+  static constexpr auto& rdinstreth = perfCounter<Hi, rdinstretImpl>;
+
+  // ----------------------------------------------------------------------
+  //
+  // RISC-V CSR Instructions
+  //
+  // ----------------------------------------------------------------------
+  struct RevZicsrInstDefaults : RevInstDefaults {
+    RevZicsrInstDefaults() {
+      SetOpcode( 0b1110011 );
+      SetRaiseFPE( true );
+      SetrdClass( RevRegClass::RegGPR );
+      Setrs2Class( RevRegClass::RegUNKNOWN );
+      Setimm( FImm );
+      SetFormat( RVTypeI );
+    }
+  };
+
+  // clang-format off
+  std::vector<RevInstEntry> ZicsrTable = {
+    { RevZicsrInstDefaults().SetMnemonic( "csrw %csr, %rs1"        ).SetFunct3( 0b001 ).SetImplFunc( csrrw      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RD( Inst ) == 0; } ) },
+    { RevZicsrInstDefaults().SetMnemonic( "csrrw %csr, %rd, %rs1"  ).SetFunct3( 0b001 ).SetImplFunc( csrrw      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RD( Inst ) != 0 && DECODE_IMM12( Inst ) != 0x1 && DECODE_IMM12( Inst ) != 0x2; } ) },
+    { RevZicsrInstDefaults().SetMnemonic( "fsflags %rs"            ).SetFunct3( 0b001 ).SetImplFunc( csrrw      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RD( Inst ) != 0 && DECODE_IMM12( Inst ) == 0x1; } ) },
+    { RevZicsrInstDefaults().SetMnemonic( "fsrm %rs"               ).SetFunct3( 0b001 ).SetImplFunc( csrrw      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RD( Inst ) != 0 && DECODE_IMM12( Inst ) == 0x2; } ) },
+
+    { RevZicsrInstDefaults().SetMnemonic( "csrrs %rd, %csr, %rs1"  ).SetFunct3( 0b010 ).SetImplFunc( csrrs      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RS1( Inst ) != 0; } ) },
+    { RevZicsrInstDefaults().SetMnemonic( "csrr %rd, %csr"         ).SetFunct3( 0b010 ).SetImplFunc( csrrs      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RS1( Inst ) == 0 && DECODE_IMM12( Inst ) != 0x1 && DECODE_IMM12( Inst ) != 0x2 && [](auto imm){ return imm < 0xc80 || imm > 0xc82; }( DECODE_IMM12( Inst ) | 0x80 ); } ) },
+    { RevZicsrInstDefaults().SetMnemonic( "frflags %rd"            ).SetFunct3( 0b010 ).SetImplFunc( csrrs      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RS1( Inst ) == 0 && DECODE_IMM12( Inst ) == 0x1; } ) },
+    { RevZicsrInstDefaults().SetMnemonic( "frrm %rd"               ).SetFunct3( 0b010 ).SetImplFunc( csrrs      ).SetPredicate( []( uint32_t Inst ){ return DECODE_RS1( Inst ) == 0 && DECODE_IMM12( Inst ) == 0x2; } ) },
+
```

Comment:
`frflags` and `frrm` are pseudoinstructions which are `csrrs` instructions underneath.